src/controller/deletecontent.js

```diff
@@ -65,31 +64,56 @@ export default function deleteContent( selection, batch, options = {} ) {
 	endPos.detach();
 }
 
+// This function is a result of reaching the Ballmer's peak for just the right amount of time.
+// Even I had troubles documenting it after a while and after reading it again I couldn't believe that it really works.
 function mergeBranches( batch, startPos, endPos ) {
-	const endPath = endPos.path;
-	const mergeEnd = Math.min( startPos.path.length - 1, endPath.length - 1 );
-	let mergeDepth = compareArrays( startPos.path, endPath );
+	const startParent = startPos.parent;
+	const endParent = endPos.parent;
 
-	if ( typeof mergeDepth == 'number' ) {
-		for ( ; mergeDepth < mergeEnd; mergeDepth++ ) {
-			const mergePath = startPos.path.slice( 0, mergeDepth );
-			mergePath.push( startPos.path[ mergeDepth ] + 1 );
+	// If both positions ended up in the same parent, then there's nothing more to merge:
+	// <$root><p>x[]</p><p>{}y</p></$root> => <$root><p>xy</p>[]{}</$root>
+	if ( startParent == endParent ) {
+		return;
+	}
 
-			const mergePos = new Position( endPos.root, mergePath );
-			const previousNode = mergePos.nodeBefore;
-			const nextNode = mergePos.nodeAfter;
+	// If one of the positions is a root, then there's nothing more to merge (at least in the current state of implementation).
+	// Theoretically in this case we could unwrap the <p>: <$root>x[]<p>[]y</p></$root>, but we don't need to support it yet
+	// so let's just abort.
+	if ( !startParent.parent || !endParent.parent ) {
+		return;
+	}
 
-			if ( !checkCanBeMerged( previousNode ) || !checkCanBeMerged( nextNode ) ) {
-				return;
-			}
+	// Check if operations we'll need to do won't need to cross object or limit boundaries.
+	// E.g., we can't merge endParent into startParent in this case:
+	// <limit><startParent>x</startParent></limit><endParent></endParent>
+	if ( !checkCanBeMerged( startParent, endParent ) ) {
+		return;
+	}
 
-			if ( nextNode.childCount > 0 ) {
-				batch.merge( mergePos );
-			} else {
-				batch.remove( nextNode );
-			}
-		}
+	// Remember next positions to merge. For example:
+	// <a><b>x[]</b></a><c><d>[]y</d></c>
+	// will become:
+	// <a><b>xy</b>[]</a><c>[]</c>
+	startPos = Position.createAfter( startParent );
+	endPos = Position.createBefore( endParent );
+
+	if ( endParent.childCount > 0 ) {
+		// At the moment, next startPos is also the position to which the endParent
+		// needs to be moved:
+		// <a><b>x[]</b></a><c><d>[]y</d></c>
+		// becomes:
+		// <a><b>x</b>[]<d>y</d></a><c>[]</c>
+		batch.move( endParent, startPos );
+
+		// To then become:
+		// <a><b>xy</b>[]</a><c>[]</c>
+		batch.merge( startPos );
+	} else {
+		batch.remove( endParent );
 	}
+
+	// Continue merging next level.
```

Comment:
I guess this example shows this behaviour: https://github.com/ckeditor/ckeditor5-engine/pull/907/files#diff-b55e8246527f4a9a00e50fb4c1c88b10R269

It makes sense, I just wonder whether the function has to be that complicated.